### PR TITLE
IOSim Timeouts API refactor and optimisation

### PR DESCRIPTION
**Pull request by bolt12**
This is the same as in https://github.com/input-output-hk/ouroboros-network/pull/3682 but on the new repository

Resolves input-output-hk/ouroboros-network#3672

### CODE REVIEW

**Comment by bolt12** [comment]
NOTE that https://github.com/input-output-hk/ouroboros-network/pull/3682 had 1 more commit that fixed a `KeepAlive` convergence test. I am going to add that as a separate PR over at `ouroboros-network`.

**Comment by coot** [comment]
Can we turn this PR into a draft, we need to do the `IOSimPOR` side before merging, isn't it?

**Comment by bolt12** [comment]
closes https://github.com/input-output-hk/ouroboros-network/issues/3672

File: io-sim/test/Test/Control/Monad/IOSimPOR/Test.hs
Comment (coot) [inline io-sim/test/Test/Control/Monad/IOSimPOR/Test.hs:8]: There are compilation warnings due to: `-Wincomplete-uni-patterns`.

**Comment by coot** [comment]
Lets consider two concurrent threads which are blocked on `threadDealy`, which expires at the same time. We don't need to race the two events because there's nothing racy on its own here, if the two threads do something later that introduces a race that race can be inverted.  Even if we deschedule the thread just after a timer delay expires (and thus we start a new step), this will not influence the ability to invert the following racing steps.

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs:400]: I have the same question, but if you look a few lines above (or in the diff before I changed stuff) when we found a suitable exception handler we call `updateRacesInSimState`. So I figured this particular case and the ones below should behave in the same way since in a nutshell we are dealing with particular `CatchFrame`s (they are not exactly `CatchFrame`s in the sense that one called catch somewhere in the code, but they are implicit catch call)

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:960]: I think the new thread _must_ start with async exceptions unmasked, rather than inheriting the masking state from the parent thread. This is important because the current locking scheme relies on being able to kill this thread, which is only possible if async exceptions are unmasked (or at least not masked uninterruptible).

See the GHC timeout code that uses `forkIOWithUnmask $ \unmask -> unmask $ do`

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:258]: So there's still a residual race here isn't there? We're now leaving our `TimeoutFrame` and we're throwing an exception to the other thread that's trying to kill us. So what can happen?

We've got two threads, each of which has as their next action to kill the other one. Due to the semantics of `throwTo` only one of these can succeed, not both (`throwTo` is in fact synchronous, it blocks until the exception is delivered).

So if this thread succeeds first, it will kill the other one, and proceed happily.

If the other thread succeeds first, it will throw a `TimeoutException` to us, but we're now outside of our `TimeoutFrame`! So things will not work as expected.

Thus I conclude that this locking scheme does not do what is intended!

Take another look at the locking scheme that GHC uses, it's not the same as what we're doing here. In the GHC timeout code, the lock is not taken when the interrupting thread is forkIO'd, but is only taken by that forkIO'd thread when it's about to kill the action thread (I'm not sure that's important, but does reduce the time interval). Furthermore, although it does look like a similar race can happen there: note that the cleanup handler uses `uninterruptibleMask_` to prevent itself being killed by the other thread. That does resolve the race.

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:994]: I'm not convinced that we need to be collecting and returning these timeout ids. I think that's due to the need to delete the locks right? If we follow the locking design I suggested then the lifetime of locks becomes simpler.

Or was it due to being strict about deleting thread delay timeouts?

**Review by dcoutts — CHANGES_REQUESTED** [review]
I've got through the first two commits. Lots of comments.

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:281]: I don't exactly remember why I went for this `ThreadDelayFrame` but I think it was so I could have access to the `TimeoutId` for that thread. remember being this case being specially tricky `timeout a (threadDelay b)` if we don't have `ThreadDelayFrame` in this case if timeout expires, when sending an exception to the `threadDelay` I wanted a way to know if that thread was under a `threadDelay` so I could remember to remove the timer

But I will revisit this issue and refactor

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:258]: GHC implementation is more subtle because it uses `registerTimeout`, `handleTimeout` and `cleanupTimeout` it also uses `tryPutMVar` and it makes it hard to translate into IOSim code, but here is my understanding. `handleTimeout` runs when the timeout expired, so it will check the lock: if it is `NotLocked` (it means that the running thread still hasn't finished, so it will fork a new thread. This new thread objective is to stop/kill the running thread since the timeout expired, **_however_** right before `throwingTo` it will try to acquire the lock, meaning it will only send the exception after immediately being able to acquire the lock. If the lock is `Locked` in `handleTimeout` it means that the running Thread won the race. Notice that that the `handleTimeout` code runs unmasked.

`cleanupTimeout` if the running thread finished and this can race with the timeout firing at the same time. `cleanupTimeout` is going to try acquire the lock, if it succeeds it just cleans the timer, otherwise it will attempt to kill the thread that is trying to kill it. `cleanupTimeout` runs its code uninterruptiblyMasked which kind of allows for the code both in `cleanupTimeout` and `handleTimeout` to solve the race with `cleanupTimeout` having a slight advantage due to be uninterruptible.

So in my code, at this point in the code, it means that the running thread finished, but we need to be careful to a potential race so we check the lock. If the lock is `Locked` it means there's already an assassin thread out there so we need to shield ourselves with an `uninterruptibleMask` and kill it. If the lock is `NotLocked` it means the timer hasn't expired and we acquire it immediately (notice that our code is sequential so not possible that the state the lock is being changed) and we just remove the timer from the queue and carry on.

So given this my code is wrong, but I just have to run my thread in an uninterruptible mask in order for it to be as faithful as GHC's implementation as possible

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:448]: This is a wrong copy paste comment, I will fix this!

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:994]: Yes, we can simplify this!

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:946]: Correct!

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:960]: Yes, good catch!

File: io-sim/src/Control/Monad/IOSim/InternalTypes.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/InternalTypes.hs:37]: Don't really know what happened! Will revert it back!

File: io-sim/src/Control/Monad/IOSim/InternalTypes.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/InternalTypes.hs:42]: Yes, this makes sense!

File: io-sim/src/Control/Monad/IOSim/Types.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Types.hs:142]: Because `timeout :: forall a. DiffTime -> IOSim s a -> IOSim s (Maybe a)` but I guess I can make it be `SimA s a` not sure there will be any difference

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs]: After the refactor this no longer is required

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs]: I don't think this is required anymore

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs]: Same

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs]: Removed!

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs]: Done!

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (coot) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs]: I would still like to hear back from John, but we don't need to be blocked on merging.

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:230]:
```suggestion
        -- There is a possible race between timeout action and the timeout expiration.
        -- We use a lock to solve the race.
```

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:250]: You called it _assassin thread_, so let's be consequent:
```suggestion
            let -- Kill the assassin thread and carry on the continuation
```

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:913]: indentation (and code formatting):
```suggestion
            return st { runqueue = runqueue',
                        threads  = threads''
                      })
```

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:843]: We also ought to record that we forked threads when we used `forkTimeoutInterruptThreads`.

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:820]: indentation:
```suggestion
        !timeoutExpired' <- traverse (\(tid, tmid, isLockedRef) -> do
                                        locked <- readSTRef isLockedRef
                                        return (tid, tmid, isLockedRef, locked)
                                     )
                                     timeoutExpired
```

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:856]: I think we can safely use `{}`, if these constructors get a new field they will not be modified here.
```suggestion
    timeoutSTMAction TimerThreadDelay {}     = return ()
    timeoutSTMAction TimerTimeout {}         = return ()
```

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:943]: indentation (we also ought to adjust to our style guide):
```suggestion
         in ( Thread { threadId      = tid'
                     , threadControl =
                        ThreadControl
                          (ThrowTo (toException (TimeoutException tmid))
                                    tid
                                    (Return ()))
                          ForkFrame
                     , threadBlocked = False
                     , threadMasking = Unmasked
                     , threadThrowTo = []
                     , threadClockId = threadClockId t
                     , threadLabel   = Just "timeout-forked-thread"
                     , threadNextTId = 1
                     }
```

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:980]: This does not need to be modified.

File: io-sim/src/Control/Monad/IOSim/Types.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Types.hs:533]:
```suggestion
    | otherwise = IOSim $ oneShot $ \k -> StartTimeout d (runIOSim action) k
```

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1015]: As in `IOSim` case, we ought to record thread started with `forkTimeoutInterruptThreads`.

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1097]: indentation & style

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1133]: idnentation & style

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1169]: This does not need to be modified.

**Review by coot — CHANGES_REQUESTED** [review]

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:855]: Shouldn't this be `EventTimerFired`?

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:858]: Should we trace `EventRegisterDelayFired` here?

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:865]: Shouldn't this be `EventThreadDelayFired`?

File: io-sim/src/Control/Monad/IOSim/Types.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Types.hs:785]: Shouldn't we also have `EventThreadDelayFired`?

File: io-sim/src/Control/Monad/IOSim/Types.hs
Comment (coot) [inline io-sim/src/Control/Monad/IOSim/Types.hs:796]: Shouldn't we have `EventTimerFired`?

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (coot) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1076]: What's strange in this block is that we log:
* `EventTimerExpired`
* `EventRegisterDelayExpired`
* `EventThreadDelayExipred`
* `EventTimeoutFired`

The last one does not fit to the other ones, so something is wrong.  Were are we tracing:
* `EventTimerFired`
* `EventRegisterDelayFired`
* `EventThreadDelayFired`

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:855]: I think I used "Expired" instead of "Fired" I can change that if you prefer

File: io-sim/src/Control/Monad/IOSim/Types.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Types.hs:785]: `EventThreadDelayExpired` is supposed to denote `EventThreadDelayFired`

File: io-sim/src/Control/Monad/IOSim/Types.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Types.hs:796]: Sam as above

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1076]: Should be here but I named them differently.

File: io-sim/src/Control/Monad/IOSim/Types.hs
Comment (coot) [reply io-sim/src/Control/Monad/IOSim/Types.hs]: ok, so it's better to conform to the same naming convention and use `EventThreadDelayFired`.

File: io-sim/test/Test/Control/Monad/IOSim/Test.hs
Comment (coot) [inline io-sim/test/Test/Control/Monad/IOSim/Test.hs:6]:
```suggestion
module Test.Control.Monad.IOSim
```
We don't need to say twice that it's a testing module :).

File: io-sim/test/Test/Control/Monad/IOSimPOR/Test.hs
Comment (coot) [inline io-sim/test/Test/Control/Monad/IOSimPOR/Test.hs:10]:
```suggestion
module Test.Control.Monad.IOSimPOR (tests) where
```

File: io-sim/test/Test/Control/Monad/Common.hs
Comment (coot) [inline io-sim/test/Test/Control/Monad/Common.hs:6]: The name is misleading as it actually contains properties & utils.  It would be better to have a better distinction between these two.

**Review by coot — APPROVED** [review]
LGTM, thanks a lot @bolt12 :)

**Pull request merged**
